frame/system/src/offchain.rs

```diff
@@ -15,353 +15,609 @@
 	}
 }
 
-/// Retrieves a public key type for given `SignAndSubmitTransaction`.
-pub type PublicOf<T, Call, X> =
-<
-	<X as SignAndSubmitTransaction<T, Call>>::CreateTransaction
-	as
-	CreateTransaction<T, <X as SignAndSubmitTransaction<T, Call>>::Extrinsic>
->::Public;
+impl<T: SigningTypes, C: AppCrypto<T::Public, T::Signature>, X> Signer<T, C, X> {
+	/// Use all available keys for signing.
+	pub fn all_accounts() -> Signer<T, C, ForAll> {
+		Default::default()
+	}
 
-/// A trait to sign and submit transactions in off-chain calls.
-///
-/// NOTE: Most likely you should not implement this trait yourself.
-/// There is an implementation for
-/// [`TransactionSubmitter`] type, which
-/// you should use.
-pub trait SignAndSubmitTransaction<T: crate::Trait, Call> {
-	/// Unchecked extrinsic type.
-	type Extrinsic: ExtrinsicT<Call=Call> + Encode;
-
-	/// A runtime-specific type to produce signed data for the extrinsic.
-	type CreateTransaction: CreateTransaction<T, Self::Extrinsic>;
-
-	/// A type used to sign transactions created using `CreateTransaction`.
-	type Signer: Signer<
-		PublicOf<T, Call, Self>,
-		<Self::CreateTransaction as CreateTransaction<T, Self::Extrinsic>>::Signature,
-	>;
-
-	/// Sign given call and submit it to the transaction pool.
+	/// Use any of the available keys for signing.
+	pub fn any_account() -> Signer<T, C, ForAny> {
+		Default::default()
+	}
+
+	/// Use provided `accounts` for signing.
 	///
-	/// Returns `Ok` if the transaction was submitted correctly
-	/// and `Err` if the key for given `id` was not found or the
-	/// transaction was rejected from the pool.
-	fn sign_and_submit(call: impl Into<Call>, public: PublicOf<T, Call, Self>) -> Result<(), ()> {
-		let call = call.into();
-		let id = public.clone().into_account();
-		let mut account = super::Account::<T>::get(&id);
-		debug::native::debug!(
-			target: "offchain",
-			"Creating signed transaction from account: {:?} (nonce: {:?})",
-			id,
-			account.nonce,
-		);
-		let (call, signature_data) = Self::CreateTransaction
-			::create_transaction::<Self::Signer>(call, public, id.clone(), account.nonce)
-			.ok_or(())?;
-		// increment the nonce. This is fine, since the code should always
-		// be running in off-chain context, so we NEVER persists data.
-		account.nonce += One::one();
-		super::Account::<T>::insert(&id, account);
-
-		let xt = Self::Extrinsic::new(call, Some(signature_data)).ok_or(())?;
-		sp_io::offchain::submit_transaction(xt.encode())
+	/// Note that not all keys will be necessarily used. The provided
+	/// vector of accounts will be intersected with the supported keys
+	/// in the keystore and the resulting list will be used for signing.
+	pub fn with_filter(mut self, accounts: Vec<T::Public>) -> Self {
+		self.accounts = Some(accounts);
+		self
+	}
+
+	/// Check if there are any keys that could be used for signing.
+	pub fn can_sign(&self) -> bool {
+		return self.accounts.is_some() && self.accounts.as_ref().unwrap_or(&vec![]).len() > 0
```

Comment:
That's not correct. We need to check if there are keys in the intersection. `self.accounts = None` is totally fine.

The simplest way to implement this logic correctly would be to use `for_any` function and just return `true` for any account that is found in the intersection (you would need to implement `for_any` for both `ForAll` and `ForAny`)

Also that reminds me that it would be good to have some tests for new traits in this pallet.